crates/matrix-sdk/tests/integration/room/joined.rs

```diff
@@ -833,3 +837,111 @@ async fn test_enable_encryption_doesnt_stay_unencrypted() {
 
     assert!(room.is_encrypted().await.unwrap());
 }
+
+#[async_test]
+async fn test_subscribe_to_requests_to_join() {
+    let server = MatrixMockServer::new().await;
+    let client = server.client_builder().build().await;
+
+    server.mock_room_state_encryption().plain().mount().await;
+
+    let room_id = room_id!("!a:b.c");
+    let f = EventFactory::new().room(room_id);
+
+    let alice_user_id = user_id!("@alice:b.c");
+    let alice_knock_event_id = event_id!("$alice-knock:b.c");
+    let alice_knock_event = f
+        .event(RoomMemberEventContent::new(MembershipState::Knock))
+        .event_id(alice_knock_event_id)
+        .sender(alice_user_id)
+        .state_key(alice_user_id)
+        .into_raw_timeline()
+        .cast();
+
+    server.mock_get_members().ok(vec![alice_knock_event]).mock_once().mount().await;
+
+    let room = server.sync_joined_room(&client, room_id).await;
+    let stream = room.subscribe_to_join_requests().await.unwrap();
+
+    pin_mut!(stream);
+
+    // We receive an initial request to join from Alice
+    let initial = assert_next_with_timeout!(stream, 100);
+    assert!(!initial.is_empty());
+
+    let alices_request_to_join = &initial[0];
+    assert_eq!(alices_request_to_join.event_id, alice_knock_event_id);
+    assert!(!alices_request_to_join.is_seen);
+
+    // We then mark the request to join as seen
+    room.mark_join_requests_as_seen(&[alice_knock_event_id.to_owned()]).await.unwrap();
+
+    // Now it's received again as seen
+    let seen = assert_next_with_timeout!(stream, 100);
+    assert!(!seen.is_empty());
+    let alices_seen_request_to_join = &seen[0];
+    assert_eq!(alices_seen_request_to_join.event_id, alice_knock_event_id);
+    assert!(alices_seen_request_to_join.is_seen);
+
+    // If we then receive a new member event for Alice that's not 'knock'
+    let alice_join_event_id = event_id!("$alice-join:b.c");
+    let joined_room_builder = JoinedRoomBuilder::new(room_id).add_state_bulk(vec![f
```

Comment:
if you use `add_state_event`, you shouldn't need the `vec![]`

Reply:
`add_state_event` only allows you to pass a `StateTestEvent` instance. I could use `StateTestEvent::Custom`, but then I need to use a hardcoded json.